File: server/lockout.ts
export class LockoutTimer {
  // A very simple, quick, inexact lockout timer
  // For each request isLocked() or updateBucketTime() must be called before addCount()
  // counts in buckets of countPeriodMs
  // each period starts with an empty bucket
  // suggest SB_LOCKOUT_TIME_MS=60000  SB_LOCKOUT_LIMIT=10
  bucketTime: number = 0;
  bucketCount: number = 0;
  bucketSize: number;
  limit: number;
  disabled: boolean;

  constructor(
    countPeriodMs: number,
    limit: number,
  ) {
    this.disabled = isNaN(countPeriodMs) || isNaN(limit) || countPeriodMs < 1 ||
      limit < 1;
    this.bucketSize = countPeriodMs;
    this.limit = limit;
  }

  updateBucketTime(): void {
    const currentBucketTime = Math.floor(Date.now() / this.bucketSize);
    if (this.bucketTime === currentBucketTime) {
      return;
    }
    // the bucket is too old - empty it
    this.bucketTime = currentBucketTime;
    this.bucketCount = 0;
  }

  isLocked(): boolean {
    if (this.disabled) {
      return false;
    }
    this.updateBucketTime();
    return this.bucketCount >= this.limit;
  }

  addCount(): void {
    // isLocked or updateBucketTime must be called first to keep bucketTime current
    this.bucketCount++;
  }
}

File: server/lockout.go
package server

import (
	"math"
	"sync"
	"time"
)

// LockoutTimer implements a simple rate limiter to prevent brute force attacks
type LockoutTimer struct {
	mutex       sync.Mutex
	bucketTime  int64
	bucketCount int
	bucketSize  int64 // duration in milliseconds
	limit       int
	disabled    bool
}

// NewLockoutTimer creates a new lockout timer
// countPeriodMs: time window in milliseconds
// limit: maximum attempts allowed in the time window
func NewLockoutTimer(countPeriodMs int, limit int) *LockoutTimer {
	disabled := math.IsNaN(float64(countPeriodMs)) || math.IsNaN(float64(limit)) ||
		countPeriodMs < 1 || limit < 1

	return &LockoutTimer{
		bucketSize: int64(countPeriodMs),
		limit:      limit,
		disabled:   disabled,
	}
}

// updateBucketTime updates the current bucket time and resets count if needed
func (lt *LockoutTimer) updateBucketTime() {
	currentBucketTime := time.Now().UnixMilli() / lt.bucketSize
	if lt.bucketTime == currentBucketTime {
		return
	}
	// the bucket is too old - empty it
	lt.bucketTime = currentBucketTime
	lt.bucketCount = 0
}

// IsLocked checks if the timer is currently locked due to too many attempts
func (lt *LockoutTimer) IsLocked() bool {
	if lt.disabled {
		return false
	}

	lt.mutex.Lock()
	defer lt.mutex.Unlock()

	lt.updateBucketTime()
	return lt.bucketCount >= lt.limit
}

// AddCount increments the attempt counter
// IsLocked() should be called first to keep bucketTime current
func (lt *LockoutTimer) AddCount() {
	if lt.disabled {
		return
	}

	lt.mutex.Lock()
	defer lt.mutex.Unlock()

	// updateBucketTime should have been called by IsLocked first
	lt.bucketCount++
}
